include/mesos/zookeeper/url.hpp

```diff
@@ -101,16 +102,35 @@ inline Try<URL> URL::parse(const std::string& url)
   // Look for the trailing '@' (if any), that's where servers starts.
   size_t index = s.find_last_of('@');
 
-  if (index != std::string::npos) {
-    return URL(s.substr(0, index), s.substr(index + 1), path);
-  } else {
+  if (index == std::string::npos)
     return URL(s, path);
+
+  std::string servers = s.substr(index + 1);
+  std::string auth = s.substr(0, index);
+
+  size_t schemeDelimiter = auth.find_first_of('!');
+
+  // If there is not '!' in URL scheme is "digest" and everything before '@' is credentials
+  std::string scheme = "digest";
+  std::string credentials = auth;
```

Comment:
Adding the authentication scheme into this location gives us a small-ish patch, but has downsides:

- The addition of an exclamation mark delimiter invalidates the URI schema, according to https://tools.ietf.org/html/rfc1034#section-3.5 .  Only letters, digits, and dashes are allowed in a "host" field.  If we (Mesos) ever change our URI parsing logic, we would need to rip out this code.
- Zookeeper's pluggable system does not seem to limit the characters in an authentication scheme (as far as I know).  If a plugin's scheme has an `@` in the name for whatever reason, the URI parsing logic here would fail spectacularly 😈 .

My recommendation is to add an optional master/agent flag which specifies the ZK authentication scheme.

Reply:
It is not in the hostname. Look:
example: zk://simple!login:password@host:port/path
hostname starts after '@'.
Everything after zk:// and before '@' has nothing to do with DNS and hostnames.

Check https://tools.ietf.org/html/rfc3986
Page 49 describes allowed delimiters in this part, explicitly including '!'.
Section 3.2.1 describes this part as "User Information".

Reply:
Ah, I missed that.  However, if we use a `!` as a delimiter, that means the user info cannot have exclamation marks.  So something like:
```
zk://user:pass!word@host:port/path
```
Would give us a "authentication scheme" of `user:pass`, which would be incorrect.

Reply:
Yes. But it is an obvious side effect of design decision to pass authentication information as URL. Which was already in place. 

It is also not a big problem. We can allow escape it somehow. Like "\!".

Reply:
Also we can choose another delimiter symbol, RFS allows to use several symbols.